### ISSUE TITLE
manual plugin does not work well with DNS wildcard auth for domain+wildcard

### ISSUE THREAD

**Issue by jvanasco** [certbot/certbot#5801]
I go into a lengthy discussion on this in the letsencrypt forum - https://community.letsencrypt.org/t/problems-flaws-with-base-wildcard-validation/57815

The basic issue is this:

When doing a manual DNS challenge for certificate that covers both a registered domain `example.com` and wildcard `*.example.com`, the order of operations for writing challenges is serial:

* client creates `_acme-challenge.example.com` for first auth
* boulder validates via dig (or whatever)
* client creates `_acme-challenge.example.com` for second auth
* boulder validates via dig (or whatever)

The problem with this approach is that the first validation tends to cache the TXT entry for `_acme-challenge.example.com` on many DNS providers for at least the duration of the TTL value.  

This type of behavior should be expected from Enterprise DNS providers - it has been an historic problem with other DNS auth systems for a decade or longer (google site ownership, identity systems, etc).

Enterprise systems don't function the same as a single BIND server - they are often fronted by a webapp/api dashboard that saves data into a RDBMS (or similar) and utilizes a readthrough cache. Often times there are internal caches or caching DNS servers between what reads the RDBMS and serves the cache.  Domain (not IP) nameservers might be geographically distributed and may be handled by roundrobin DNS. For example, one provider has `ns1.example.com` through `ns5.example.com`, each one in a different geographic region from the internal system that services the write; another provider might have `ns1.example.com` resolve to multiple IP addresses in a roundrobin style.  Large systems also often use internal caching and replication where the public IP is just the gateway for a cluster of nameservers.  Few systems offer a DNS flush at all on updates. Few state they handle propagation via an instant `push` or write-through mechanism (vs a read-through system).  These systems (rightly) appear to cache on just the key, because caching on a key+value combo makes no sense at all. 

Suggested ways to handle this are

1. Make manual DNS verification write both challenges parallel (or in serial) before any validation happens.
2. Use a single auth challenge for root+wildcard
3. Use a separate DNS key for wildcards

The current approach requires me to `sleep` at twice the TTL on several DNS services in order for the second challenge to validate (suggesting internal cache systems that respect TTL).  One vendor I tested with has a 5minute minimum TTL, which makes it largely incompatible with this auth mechanism.

Edit: Referencing this to a similar feature on https://github.com/certbot/certbot/pull/5712

**Comment by sydneyli** [comment]
Thanks for the in-depth report! (2) and (3) require changes to how boulder/ACME do validation, but (1) --performing the challenges in parallel before the first validation happens-- is manageable. Will take a look later today.

**Comment by sydneyli** [comment]
@jvanasco I took a look, and it seems like this is already how the manual authenticator works! It doesn't ask LE to verify until all your DNS records are set. The remaining suggestions are changes that must be made against ACME/boulder.

Closing this issue.

**Comment by jvanasco** [comment]
@sydneyli thanks for looking. would you mind showing me the relevant files where this happens? I want to dig in myself and see if this is a version issue or something odd with my os.

**Comment by sydneyli** [comment]
@jvanasco Most of the relevant logic is in the [auth_handler](https://github.com/certbot/certbot/blob/master/certbot/auth_handler.py#L73) -- `_solve_challenges` calls `auth.perform`. In manual DNS mode, this asks the user to set all the relevant TXT records, and is defined [here](https://github.com/certbot/certbot/blob/master/certbot/plugins/manual.py#L139). Then [`_respond`](https://github.com/certbot/certbot/blob/master/certbot/auth_handler.py#L80) pings the ACME endpoint.

Feel free to re-open this issue if yours persists, or if you find anything funky while digging!

**Comment by jvanasco** [comment]
@sydneyli Thanks a ton! Sorry I didn't pick up on that sooner.  I'll run some tests and try to figure out WTF is happening.

**Comment by sydneyli** [comment]
@jvanasco No problem! Let us know if you have more qs about the codebase.

**Comment by jvanasco** [comment]
Thanks a ton. Sorry for annoying you with this.

I did a whole bunch of debuging and tracing, and it seems:

1. I read the output / logs VERY wrong before.
2. The caching issues are elsewhere on the DNS systems. They probably push on first record, it gets cached, then don't pull it for subsequent records.  Customer support on these companies are often a black box, and it takes a while to get a real answer.  I miss the days when I actually knew the engineers at my vendors.

I can probably work around this with a simple script, and have some ideas for the manual plugin I may suggest in a PR

**Comment by sydneyli** [comment]
no worries at all. since this sounds like a different concern than the one originally raised, I'd suggest making a new issue detailing the changes/improvements you'd like to make before submitting a PR!

**Comment by jvanasco** [comment]
@mcblum the best way to handle this is migrate to running your own acme-dns instance. it's not that hard to get going, but is a bit of a pain.  my current interim fix has been to patch the `perform` method of the Manual plugin to do a quick sleep once the setup is done (with over 40 domains managed, a 3 hour validation routine that might fail is not an option for me).

In #5805, I suggested the environment vars be extended to track the batch index - that would let you discern the last authentication via testing in shell script with `"$(($CERTBOT_BATCH_LEN <= ($CERTBOT_BATCH_IDX + 1)))"`.  I link to working fork/commit with this code, which tested fine for me against the production servers.

I believe the key of the txt entry is controlled by the ACME-spec and implemented by LetsEncrypt/boulder, so certbot can't influence it (though some people work on both).

**Comment by mcblum** [comment]
@jvanasco I'm not even sure what an `acme-dns` instance is :) I think I'll have to do some digging on this. It's interesting that the key is part of the spec, every validation I've ever done from, let's say, Google has generated a random string for the key and a random string for the value. I will also admit that I don't completely know how all of this works :)

Sounds like what you're doing right now is running the script, setting the DNS record, sleeping for the same amount of time it takes to invalidate the cache of the dns record, rinse and repeat for all domains. Is that correct?

**Comment by jvanasco** [comment]
@mcblum  acme-dns is an API driven DNS server, written specifically for validating the acme DNS-01 challenges ( https://github.com/joohoi/acme-dns ).  You basically run it on a machine you control, and CNAME the `_acme-challenge.example.com` record onto it. (there are more instructions on that project's page, and there is a more detailed documentation PR that is being worked on right now).  There is a certbot plugin that automates it all for you, and you only run the acme-dns server when you are doing a certificate renewal/issue.  Using it gets around the cache issues AND the various security issues (Most DNS API credentials are too powerful to leave on a web server for renewal).

In terms of dealing with your current nameservers...

If you only have a few domains, you can just sleep for 2-5mins in the manual hook script *after* setting every challenge.  I use a hook script that sets the values using `lexicon` (the approach is documented on this great blog post https://id-rsa.pub/post/certbot-auto-dns-validation-with-lexicon/).

But if you have a lot of domains, that approach doesn't work well because each domain needs 2 challenges.  Using the fastest cloud DNS system I have access to as an example, 43 domains == 86 challenges; but there is a 120s minimum sleep needed after the last challenge for it to be available, so it would take me *at least* 2hrs 52 mins to find out if it worked or not.

So... my setup for the manual plugin is this:
* A hook script is used to set the challenge.  Certbot loops setting all the challenges on the same certificate.  (I don't know how it handles multiple certs in a renewal)
* After setting the last challenge, I sleep for the amount of time it takes to invalidate the longest cache that is applicable to the certs i need.  
* With a 60s TTL on Namecheap, my minimum wait is about 120s.  (With linode and dreamhost, it is about 5 minutes).  To process those 43 domains using this approach, I only have to sleep the minimum wait time per certificate -- so the difference between issuing everything on one certificate goes from "about 3 hours" to "about 5 minutes".
